cbounter/ht_common.c

```diff
@@ -0,0 +1,963 @@
+    #endif
+}
+
+static PyTypeObject HT_VARIANT(_ITER_TYPE_Type) = {
+    #if PY_MAJOR_VERSION >= 3
+    PyVarObject_HEAD_INIT(NULL, 0)
+    #else
+    PyObject_HEAD_INIT(NULL)
+    0,                               /* ob_size */
+    #endif
+    "htc." HT_TYPE_STRING "_iter",             /*tp_name*/
+    sizeof(HT_VARIANT(_ITER_TYPE)), /*tp_basicsize*/
+    0,                         /*tp_itemsize*/
+    (destructor)HT_VARIANT(_ITER_dealloc), /*tp_dealloc*/
+    0,                         /*tp_print*/
+    0,                         /*tp_getattr*/
+    0,                         /*tp_setattr*/
+    0,                         /*tp_compare*/
+    0,                         /*tp_repr*/
+    0,                         /*tp_as_number*/
+    0,                         /*tp_as_sequence*/
+    0,                         /*tp_as_mapping*/
+    0,                         /*tp_hash */
+    0,                         /*tp_call*/
+    0,                         /*tp_str*/
+    0,                         /*tp_getattro*/
+    0,                         /*tp_setattro*/
+    0,                         /*tp_as_buffer*/
+    #if PY_MAJOR_VERSION >= 3
+    Py_TPFLAGS_DEFAULT,
+    #else
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_ITER,
+      /* tp_flags: Py_TPFLAGS_HAVE_ITER tells python to
+         use tp_iter and tp_iternext fields. */
+    #endif
+    "Internal HTC iterator object.",           /* tp_doc */
+    0,  /* tp_traverse */
+    0,  /* tp_clear */
+    0,  /* tp_richcompare */
+    0,  /* tp_weaklistoffset */
+    HT_VARIANT(_ITER_iter),  /* tp_iter: __iter__() method */
+    HT_VARIANT(_ITER_iternext)  /* tp_iternext: next() method */
+};
+
+static inline HT_VARIANT(_ITER_TYPE) *
+HT_VARIANT(_make_iterator)(HT_TYPE *self, char result_type)
+{
+    HT_VARIANT(_ITER_TYPE) * iterator = PyObject_New(HT_VARIANT(_ITER_TYPE), &HT_VARIANT(_ITER_TYPE_Type));
+    if (!iterator)
+        return NULL;
+
+    if (!PyObject_Init((PyObject *)iterator, &HT_VARIANT(_ITER_TYPE_Type))) {
+        Py_DECREF(iterator);
+        return NULL;
+    }
+
+    iterator->i = 0;
+    iterator->hashtable = self;
+    iterator->result_type = result_type;
+    Py_INCREF(self);
+    return iterator;
+}
+
+static HT_VARIANT(_ITER_TYPE) *
+HT_VARIANT(_HT_iter_KV)(HT_TYPE *self)
+{
+    return HT_VARIANT(_make_iterator)(self, ITER_RESULT_KV_PAIRS);
+}
+
+static HT_VARIANT(_ITER_TYPE) *
+HT_VARIANT(_HT_iter_K)(HT_TYPE *self)
+{
+    return HT_VARIANT(_make_iterator)(self, ITER_RESULT_KEYS);
+}
+
+
+static PyMethodDef HT_VARIANT(_methods)[] = {
+    {"increment", (PyCFunction)HT_VARIANT(_increment), METH_VARARGS,
+     "Adds a string to the counter."
+    },
+    {"cardinality", (PyCFunction)HT_VARIANT(_cardinality), METH_NOARGS,
+     "Returns an estimate for the number of distinct items inserted into the counter. Does not work correctly when values are deleted!"
+    },
+    {"total", (PyCFunction)HT_VARIANT(_total), METH_NOARGS,
+     "Returns sum of all counts in the counter."
```

Comment:
Please add a note explaining how this is different from `sum(values)` and what are the assumptions for the method to work correctly.

Also, consider renaming to `sum`, which has clearer semantics IMO. With `total`, I don't know whether that's the sum of values or the number of keys.